proposals/patterns3.md

```diff
@@ -0,0 +1,253 @@
+```csharp
+if ((e1, e2) is (0, int i) or (int i, 0))
+{
+    M(i);
+}
+```
+
+Here, the variable `i` is definitely assigned inside the block, and takes it value from the other element of the tuple when a zero element is found.
+
+It has also been suggested to permit variables to be (multiply) defined in every case of a case block:
+
+```csharp
+    case (0, int x):
+    case (int x, 0):
+        Console.WriteLine(x);
+```
+
+To make any of this work, we would have to carefully define where such multiple definitions are permitted and under what conditions such a variable is considered definitely assigned.
+
+Should we elect to defer such work until later (which I advise), we could say in C# 9
+- beneath a `not` or `or`, pattern variables may not be declared.
+
+Then, we would have time to develop some experience that would provide insight into the possible value of relaxing that later.
+
+### Diagnostics, subsumption, and exhaustiveness
+
+These new pattern forms introduce many new opportunities for diagnosable programmer error.  We will need to decide what kinds of errors we will diagnose, and how to do so.  Here are some examples:
+
+``` csharp
+case >= 0 and <= 100D:
+```
+
+This case can never match (because the input cannot be both an `int` and a `double`).  We already have an error when we detect a case that can never match, but its wording ("The switch case has already been handled by a previous case" and "The pattern has already been handled by a previous arm of the switch expression") may be misleading in new scenarios.  We may have to modify the wording to just say that the pattern will never match the input.
+
+``` csharp
+case 1 and 2:
+```
+
+Similarly, this would be an error because a value cannot be both `1` and `2`.
+
+``` csharp
+case 1 or 2 or 3 or 1:
+```
+
+This case is possible to match, but the `or 1` at the end adds no meaning to the pattern.  I suggest we should aim to produce an error whenever some conjunct or disjunct of a compound pattern does not either define a pattern variable or affect the set of matched values.
+
+``` csharp
+case < 2: break;
+case 0 or 1 or 2 or 3 or 4 or 5: break;
+```
+
+Here, `0 or 1 or` adds nothing to the second case, as those values would have been handled by the first case.  This too deserves an error.
+
+``` csharp
+byte b = ...;
+int x = b switch { <100 => 0, 100 => 1, 101 => 2, >101 => 3 };
+```
+
+A switch expression such as this should be considered *exhaustive* (it handles all possible input values).
+
+In C# 8.0, a switch expression with an input of type `byte` is only considered exhaustive if it contains a final arm whose pattern matches everything (a *discard-pattern* or *var-pattern*).  Even a switch expression that has an arm for every distinct `byte` value is not considered exhaustive in C# 8.  In order to properly handle exhaustiveness of relational patterns, we will have to handle this case too.  This will technically be a breaking change, but one no user is likely to notice.
```

Comment:
```suggestion
In C# 8.0, a switch expression with an input of type `byte` is only considered exhaustive if it contains a final arm whose pattern matches everything (a *discard-pattern* or *var-pattern*).  Even a switch expression that has an arm for every distinct `byte` value is not considered exhaustive in C# 8.  In order to properly handle exhaustiveness of relational patterns, we will have to handle this case too.  This will technically be a breaking change, but no user is likely to notice.
```

Reply:
It sounds like you're quoting this paragraph verbatim. Is that what you intended?

---
In reply to: [409223574](https://github.com/dotnet/csharplang/pull/3361#discussion_r409223574) [](ancestors = 409223574)